### PR TITLE
Feature #116 - Provide better names for the test plan specification

### PR DESCRIPTION
**Pull request by bechte**
I will need to rebase it onto master, but don't have time right now. Still, I wanted to provide it to you guys for review purposes. Therefore, I've already created this pull request.

### CODE REVIEW

File: junit4-runner/src/main/java/org/junit/gen5/junit4/runner/JUnit5.java [hunk 1]
```diff
@@ -131,38 +130,29 @@ private void addFiltersFromAnnotations(DiscoveryRequest spec) {
 	}
 
 	private List<DiscoverySelector> getPackageSpecificationElements() {
-		return stream(getPackageNames()).map(PackageTestPlanSpecificationElementBuilder::byPackageName).collect(
-			toList());
+		return stream(getPackageNames()).map(PackageSelectorBuilder::byPackageName).collect(toList());
```
Comment (jlink) [inline junit4-runner/src/main/java/org/junit/gen5/junit4/runner/JUnit5.java]: Not sure why JUnit5 should use the builders from dsl package. This creates an extra dependency - especially if the dsl should move to a seperate project - without simplifying the code.
`PackageSelector::new` would be even shorter than `PackageSelectorBuilder::byPackageName`

And - what's more important - the dsl is supposed to be less stable than the underlying abstractions.

**Comment by jlink** [comment]
I've been thinking about the DSL. I think we have two different scenarios in which hiding the raw API (DiscoveryRequest etc.)  might make sense:

1. Allowing easy formulation of Requests _by people_ (e.g.) in test cases or in a yet to come suite definition. This is what the current dsl package aims at and where a fluent builder API makes sense.

2. Shielding IDE and tool builders from evolutionary changes in our structure and methods. This would require some kind of factory which can be comfortably used _programmatically_. The design patterns behind should probably go more into the direction of factories. The most important aspect of this factory or factories would be _stability_.

What do the others think?

**Comment by marcphilipp** [comment]
I've rebased this branch in #120.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/FilterResult.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	private final boolean filtered;
+
+	private final Optional<String> reason;
+
+	private FilterResult(boolean filtered, String reason) {
+		this.filtered = filtered;
+		this.reason = Optional.ofNullable(reason);
+	}
+
+	public boolean isFiltered() {
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/FilterResult.java]: Whether something is filtered has no clear semantics IMHO. I think we should change it into `isAccepted()` or something similar.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java [hunk 3]
```diff
@@ -24,65 +23,77 @@
...
 public final class DiscoveryRequest {
-	private final List<DiscoverySelector> elements = new LinkedList<>();
+	// Selectors provided to the engines to be used for finding tests
+	private final List<DiscoverySelector> selectors = new LinkedList<>();
+
+	// Filter based on the engine id
+	private final List<EngineIdFilter> engineIdFilters = new LinkedList<>();
+
+	// Discovery filters are handed through to all test engines to be applied during discovery
+	private final List<DiscoveryFilter> discoveryFilters = new LinkedList<>();
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java]: This is a raw type, it should be `List<DiscoveryFilter<?>>`.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java [hunk 4]
```diff
@@ -24,65 +23,77 @@
...
 
-	public void addElements(Collection<DiscoverySelector> elements) {
-		elements.forEach(this::addElement);
+	public void addEngineIdFilters(Collection<EngineIdFilter> engineIdFilters) {
+		this.engineIdFilters.addAll(engineIdFilters);
 	}
 
-	public void addEngineFilter(EngineFilter engineFilter) {
-		this.engineFilters.add(engineFilter);
+	public void addFilter(DiscoveryFilter discoveryFilter) {
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java]: `DiscoveryFilter<?>`

File: junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java [hunk 5]
```diff
@@ -24,65 +23,77 @@
...
+		this.discoveryFilters.add(discoveryFilter);
 	}
 
-	public void addEngineFilters(Collection<EngineFilter> engineFilters) {
-		this.engineFilters.addAll(engineFilters);
+	public void addFilters(Collection<DiscoveryFilter> discoveryFilters) {
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java]: `Collection<DiscoveryFilter<?>>`

File: junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java [hunk 6]
```diff
@@ -24,65 +23,77 @@
...
 
-	public List<EngineFilter> getEngineFilters() {
-		return unmodifiableList(this.engineFilters);
+	public List<EngineIdFilter> getEngineIdFilters() {
+		return unmodifiableList(this.engineIdFilters);
 	}
 
-	public <T extends EngineFilter> List<T> getEngineFiltersByType(Class<T> filterType) {
-		return this.engineFilters.stream().filter(filterType::isInstance).map(filterType::cast).collect(toList());
+	public <T extends DiscoveryFilter> List<T> getFilterByType(Class<T> filterType) {
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/DiscoveryRequest.java]: `<T extends DiscoveryFilter<?>>`

File: junit-engine-api/src/main/java/org/junit/gen5/engine/specification/dsl/DiscoveryRequestBuilder.java [hunk 7]
```diff
@@ -27,34 +25,31 @@
...
+ *       byTestMethod(TestDescriptorTests.class, "test1"),
+ *       byTestMethod(TestDescriptorTests.class, "testWithParams", ParameterType.class),
+ *       byTestMethod(TestDescriptorTests.class, testMethod),
+ *       byPath("/my/local/path1"),
+ *       byPath("/my/local/path2"),
+ *       byUniqueId("unique-id-1"),
+ *       byUniqueId("unique-id-2")
  *     )
+ *     .filterBy(engineIds("junit5"))
+ *     .filterBy(classNamePattern("org.junit.gen5.tests"), classNamePattern("org.junit.sample"))
```
Comment (marcphilipp) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/specification/dsl/DiscoveryRequestBuilder.java]: Are those valid patterns? I think they should end with `.*`.

**Comment by marcphilipp** [comment]
I've added a few comments inline on this pull request. However, I think it would be better to continue to work on this based on #120.

File: junit4-runner/src/main/java/org/junit/gen5/junit4/runner/JUnit5.java (hunk 1 above)
Comment (bechte) [inline junit4-runner/src/main/java/org/junit/gen5/junit4/runner/JUnit5.java]: True. I didn't introduce this intentionally. Especially, as we wanted to remove the builder in the next step. I will change this line.

File: junit-engine-api/src/main/java/org/junit/gen5/engine/FilterResult.java (hunk 2 above)
Comment (bechte) [inline junit-engine-api/src/main/java/org/junit/gen5/engine/FilterResult.java]: That's fine with me. This also is closer to the semantics of Preconditions, therefore, I think it's really better. I will change it as proposed.

**Comment by bechte** [comment]
Concerning the DSL: I would continue on our team decision and remove it, after we have the changes in place in master and provide it as a separate package. I really think we could serve both of your points in one DSL, as the current version comes with factories. But I really think we should discuss this again in the team. Maybe monday?

**Comment by jlink** [comment]
Monday in person or Tuesday on Skype :)

2016-01-16 22:00 GMT+01:00 Stefan Bechtold <notifications@github.com>:

**Comment by marcphilipp** [comment]
:+1:

**Comment by bechte** [comment]
This pull request will not be merged. The changes have been merged in #120. I will keep this request open until I have made the corrections mentioned here (in the afternoon).

**Comment by bechte** [comment]
Changes have been made in master, see 8442785b.

**Pull request closed without merging**